File: 26_Chomp/javascript/chomp.js
// CHOMP
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

var a = [];
var r;
var c;

function init_board()
{
    for (i = 1; i <= r; i++)
        for (j = 1; j <= c; j++)
            a[i][j] = 1;
    a[1][1] = -1;
}

function show_board()
{
    print("\n");
    print(tab(7) + "1 2 3 4 5 6 7 8 9\n");
    for (i = 1; i <= r; i++) {
        str = i + tab(6);
        for (j = 1; j <= c; j++) {
            if (a[i][j] == -1)
                str += "P ";
            else if (a[i][j] == 0)
                break;
            else
                str += "* ";
        }
        print(str + "\n");
    }
    print("\n");
}

// Main program
async function main()
{
    print(tab(33) + "CHOMP\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    for (i = 1; i <= 10; i++)
        a[i] = [];
    // *** THE GAME OF CHOMP *** COPYRIGHT PCC 1973 ***
    print("\n");
    print("THIS IS THE GAME OF CHOMP (SCIENTIFIC AMERICAN, JAN 1973)\n");
    print("DO YOU WANT THE RULES (1=YES, 0=NO!)");
    r = parseInt(await input());
    if (r != 0) {
        f = 1;
        r = 5;
        c = 7;
        print("CHOMP IS FOR 1 OR MORE PLAYERS (HUMANS ONLY).\n");
        print("\n");
        print("HERE'S HOW A BOARD LOOKS (THIS ONE IS 5 BY 7):\n");
        init_board();
        show_board();
        print("\n");
        print("THE BOARD IS A BIG COOKIE - R ROWS HIGH AND C COLUMNS\n");
        print("WIDE. YOU INPUT R AND C AT THE START. IN THE UPPER LEFT\n");
        print("CORNER OF THE COOKIE IS A POISON SQUARE (P). THE ONE WHO\n");
        print("CHOMPS THE POISON SQUARE LOSES. TO TAKE A CHOMP, TYPE THE\n");
        print("ROW AND COLUMN OF ONE OF THE SQUARES ON THE COOKIE.\n");
        print("ALL OF THE SQUARES BELOW AND TO THE RIGHT OF THAT SQUARE\n");
        print("INCLUDING THAT SQUARE, TOO) DISAPPEAR -- CHOMP!!\n");
        print("NO FAIR CHOMPING SQUARES THAT HAVE ALREADY BEEN CHOMPED,\n");
        print("OR THAT ARE OUTSIDE THE ORIGINAL DIMENSIONS OF THE COOKIE.\n");
        print("\n");
    }
    while (1) {
        print("HERE WE GO...\n");
        f = 0;
        for (i = 1; i <= 10; i++) {
            a[i] = [];
            for (j = 1; j <= 10; j++) {
                a[i][j] = 0;
            }
        }
        print("\n");
        print("HOW MANY PLAYERS");
        p = parseInt(await input());
        i1 = 0;
        while (1) {
            print("HOW MANY ROWS");
            r = parseInt(await input());
            if (r <= 9)
                break;
            print("TOO MANY ROWS (9 IS MAXIMUM). NOW ");
        }
        while (1) {
            print("HOW MANY COLUMNS");
            c = parseInt(await input());
            if (c <= 9)
                break;
            print("TOO MANY COLUMNS (9 IS MAXIMUM). NOW ");
        }
        print("\n");
        init_board();
        while (1) {
            // Print the board
            show_board();
            // Get chomps for each player in turn
            i1++;
            p1 = i1 - Math.floor(i1 / p) * p;
            if (p1 == 0)
                p1 = p;
            while (1) {
                print("PLAYER " + p1 + "\n");
                print("COORDINATES OF CHOMP (ROW,COLUMN)");
                str = await input();
                r1 = parseInt(str);
                c1 = parseInt(str.substr(str.indexOf(",") + 1));
                if (r1 >= 1 && r1 <= r && c1 >= 1 && c1 <= c && a[r1][c1] != 0)
                    break;
                print("NO FAIR. YOU'RE TRYING TO CHOMP ON EMPTY SPACE!\n");
            }
            if (a[r1][c1] == -1)
                break;
            for (i = r1; i <= r; i++)
                for (j = c1; j <= c; j++)
                    a[i][j] = 0;
        }
        // End of game detected
        print("YOU LOSE, PLAYER " + p1 + "\n");
        print("\n");
        print("AGAIN (1=YES, 0=NO!)");
        r = parseInt(await input());
        if (r != 1)
            break;
    }
}

main();

File: 26_Chomp/perl/chomp.pl
#!/usr/bin/perl


my @cookie;

&main;

sub main {
	my $answer = 1;
	until ($answer == 0) {
		$answer=&game_play;
	}
}

sub game_play {
	
	# the initial game set up
	&print_intro;
	my ($players,$rows,$cols) = &get_user_info;
	&create_gameboard($rows,$cols);
	&print_gameboard($rows,$cols);
	my $game_over = 0;
	my $player = 0;

	# continuous loop until the poison pill is swallowed
	until ($game_over == 1) {
		if ($player > ($players-1)) {		#checks to make sure we're just looping thru valid players 
			$player = 0;
		}
		$player++;
		my ($user_row,$user_col) = &get_player_row_col($player,$rows,$cols);
		if ($cookie[$user_row][$user_col] == -1) {
			print "YOU LOSE, PLAYER $player\n\n";
			print "AGAIN (1=YES, 0=NO!)\n";
			my $answer=<STDIN>;
			chomp($answer);
			return($answer);
		}
		&modify_gameboard($rows,$cols,$user_row,$user_col);
		&print_gameboard($rows,$cols);
	}
	
}

sub get_player_row_col {
	my ($player,$row,$col) = @_;
	my @coords;
	my $validity="invalid";
	# Getting coordinates from user
	until ($validity eq "valid") {
		print "PLAYER $player COORDINATES OF CHOMP (ROW,COLUMN)\n";
		my $input=<STDIN>;
		chomp($input);
		@coords = split/,/,$input;

		#verifying coordinates are valid
		if ($coords[0] < 1 || $coords[0] > $row || $coords[1] < 1 || $coords[1] > $col || $cookie[$coords[0]][$coords[1]] == 0) {
			print "NO FAIR. YOU'RE TRYING TO CHOMP ON EMPTY SPACE!\n";
		}
		else {
			$validity="valid";
		}
	}
	return($coords[0],$coords[1]);
}

sub get_user_info {
	my ($players,$rows,$cols)=0;
	until ($players > 0) {
		print "HOW MANY PLAYERS\n";
		$players=<STDIN>;
		chomp($players);
	}
	until ($rows > 0 && $rows < 10) {
		print "HOW MANY ROWS\n";
		$rows=<STDIN>;
		chomp($rows);
		if ($rows > 9) {
			print "TOO MANY ROWS (9 IS MAXIMUM). NOW, ";
		}
	}
	until ($cols > 0 && $cols < 10) {
		print "HOW MANY COLUMNS\n";
		$cols=<STDIN>;
		chomp($cols);
		if ($cols > 9) {
			print "TOO MANY COLUMNS (9 IS MAXIMUM). NOW, ";
		}
	}
	return($players,$rows,$cols);
}

sub print_intro{
	print ' ' x 33 . "CHOMP\n";
	print ' ' x 15 . "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n\n";
	print "THIS IS THE GAME OF CHOMP (SCIENTIFIC AMERICAN, JAN 1973)\n";
	print "DO YOU WANT THE RULES (1=YES, 0=NO!)";
	my $answer = <STDIN>;
	chomp($answer);
	if ($answer == 0) {
		return;
	}
	else {
	 	print "CHOMP IS FOR 1 OR MORE PLAYERS (HUMANS ONLY).\n\n";
	 	print "HERE'S HOW A BOARD LOOKS (THIS ONE IS 5 BY 7):\n";
		&create_gameboard(5,7);
		&print_gameboard(5,7);
	 	print "THE BOARD IS A BIG COOKIE - R ROWS HIGH AND C COLUMNS\n";
	 	print "WIDE. YOU INPUT R AND C AT THE START. IN THE UPPER LEFT\n";
	 	print "CORNER OF THE COOKIE IS A POISON SQUARE (P). THE ONE WHO\n";
	 	print "CHOMPS THE POISON SQUARE LOSES. TO TAKE A CHOMP, TYPE THE\n";
	 	print "ROW AND COLUMN OF ONE OF THE SQUARES ON THE COOKIE.\n";
	 	print "ALL OF THE SQUARES BELOW AND TO THE RIGHT OF THAT SQUARE\n";
	 	print "(INCLUDING THAT SQUARE, TOO) DISAPPEAR -- CHOMP!!\n";
	 	print "NO FAIR CHOMPING SQUARES THAT HAVE ALREADY BEEN CHOMPED,\n";
	 	print "OR THAT ARE OUTSIDE THE ORIGINAL DIMENSIONS OF THE COOKIE.\n\n";
	 	print "HERE WE GO...\n";
		undef @cookie;
	}
}

#initial creation of the gameboard
sub create_gameboard {
	my $rows = shift;
	my $cols = shift;
	foreach my $row (1..($rows)) {
		foreach my $col (1..($cols)) {
			$cookie[$row][$col]=1;
		}
	}
	$cookie[1][1]=-1;
}

#modification of the gameboard based on the input from the player
sub modify_gameboard {
	my ($rows,$cols,$user_row,$user_col) = @_;
	foreach my $row ($user_row..($rows)) {
		foreach my $col ($user_col..($cols)) {
			$cookie[$row][$col]="  ";	
		}
	}
}

#prints the gameboard based on the current state of the gameboard
sub print_gameboard {
	my ($rows,$cols) = @_;
	foreach my $col (1..$cols) {
		if ($col == $cols) {
			print "$col\n";
		}
		elsif ($col == 1) {
			print "\t  $col ";
		}
		else {
			print "$col ";
		}
	}
	foreach my $row (1..($rows)) {
		print "\t$row ";
		foreach my $col (1..($cols)) {
			if ($cookie[$row][$col] == 1) {
				print "* ";
			}
			if ($cookie[$row][$col] == 0) {
				print "  ";
			}
			if ($cookie[$row][$col] == -1) {
				print "P ";
			}
		}
		print "\n";
	}
}
